### PR TITLE
Display /tags in alphabetical order

### LINKED ISSUE #240: Display tags in alphabetical order.
**Is your feature request related to a problem? Please describe.**
The `/tags` command currently displays tags in an order that is not alphabetical, maybe random, maybe based on date added.
This makes it difficult for users to find a tag.
**Describe the solution you'd like**
We display the tags in an alphabetical order.

**Additional context**
image

### PR DESCRIPTION
**Pull request by IslamSakrak**
Fixes and closes #240 by displaying each tag in a separate line as bullet points and sorting them alphabetically.

Also, added a warn-log message if there are more than 200 tags in order to make us aware early enough of approaching the "line limit" (actually a character limit) of about 300 tags (~4k characters).

example

---
## Checklist

* [x] Wait for #295 to be merged

### CODE REVIEW

**Comment by Zabuzard** [comment]
Could you share a screenshot of how it looks like now in Discord? Then its easier to understand what you did.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:49]: * `List` not `ArrayList`
* `list` is a bad method name. It should be named according to what it contains, `tags` or `tagIds` for example.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:48]: Why the empty line here?

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:51]: `200` is a magic value and should be replaced by a proper constant with a decent name instead. for example `MAX_TAGS_WARNING_THRESHOLD`

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:53]: You should initialize the logger as static field during startup and not like that, see the other classes.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:55]: this is definitely an error and wrong. dont change the log level.

**Comment by marko-radosavljevic** [comment]
Also, don't forget to do `./gradlew spotlessApply`. ^^

Thanks for contributing to our project. :relaxed: :heart:

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:56]: * dont use capslock
* dont say WARNING, it is already a warning
* the message only explains the code condition but instead it should explain what the actual problem is and why this message is here in the first place...

for example
```
The amount of tags is very high and might soon exceed the maximum character limit. The code should be adjusted to support this edge case soon.
```
together with a `// TODO Implement this edge case` next to it, so that we do not forget to track it.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:61]: * Dont stuff this into the reply. It is really hard to read. Create the message string before and put it into a variable instead
* Why are you joining multiple times, this looks very confusing and is probably also wrong. The outer join joins only a single string, thats definitely wrong.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:49]: That said, why are you even creating a new list? Just work with what `getAllIds()` returns directly.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (IslamSakrak) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:51]: > Could you share a screenshot of how it looks like now in Discord? Then its easier to understand what you did.

image

**Comment by IslamSakrak** [comment]
> Could you share a screenshot of how it looks like now in Discord? Then its easier to understand what you did.

image

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (IslamSakrak) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:49]: Yes I agree with you,  il find another way of doing it  👍

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (IslamSakrak) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:48]: Though it would be more readable if i create some empty space

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (IslamSakrak) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:51]: Sounds good!

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (IslamSakrak) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:56]: Thank you! Il update and make a PR

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (IslamSakrak) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:61]: I found my self in some problem that, 

in 
`String.join("\n ",list.stream().sorted().collect(Collectors.joining("\n * "))` 

it would show the `*` bullet point in all elements except the first one, that is why I had to add `"* "` before the string join

But Il do as you said and remove what is redundant to make it cleaner 👍

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:48]: Fair. Up to you. Its just unusual to have an empty line at the beginning of a method. Never seen that before.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:47]: Constants are usually `private static final` fields in the class. Id suggest you change that. Otherwise, looks good.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:48]: Now you call `tagSystem.getAllIds()` twice. But the call is potentially expensive (it reads a database). Call it only once and put it into a variable.

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:52]: Dont add a `\n`. It ensures that stuff on its own.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (IslamSakrak) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:52]: Must have slipped when I copied the feedback message

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:57]: Actually, I propose a minor change for better UX. Replace the `*` by [U+2022 bullet point](https://www.compart.com/en/unicode/U+2022), i.e. `•`

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:16]: unused import? remove if so

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:36]: NIT: usually static stuff comes before non static stuff. if you want, you could reorder this, i.e. `tagSystem` below the other two. ur choice though.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (IslamSakrak) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java]: fixed

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (illuminator3) [inline application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:35]: Put a space here to separate the static and non-static fields.

**Review by illuminator3 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java
Comment (IslamSakrak) [reply application/src/main/java/org/togetherjava/tjbot/commands/tags/TagsCommand.java:37]: Done

**Review by Zabuzard — DISMISSED** [review]

**Comment by Zabuzard** [comment]
CodeFactor complains about the usage of the unicode usage and wants us to use the bullet point directly instead, it says:

> You should not make code less readable simply out of fear that some programs might not handle non-ASCII characters properly. If that should happen, those programs are broken and they should be fixed.

I kinda agree but I am unsure. The fact that you already struggled with the correct UTF-8 setup shows me that it might be tricky for beginners and would just introduce more annoyance for some people.

@marko-radosavljevic opinions?

**Comment by Tais993** [comment]
> The fact that you already struggled with the correct UTF-8 setup

If we just set the UTF in the build.gardle no-one will struggle with it?

**Comment by IslamSakrak** [comment]
> CodeFactor complains about the usage of the unicode usage `\u2022` and wants us to use the bullet point directly instead `•`, it says:
> 
> > You should not make code less readable simply out of fear that some programs might not handle non-ASCII characters properly. If that should happen, those programs are broken and they should be fixed.
> 
> I kinda agree but I am unsure. The fact that you already struggled with the correct UTF-8 setup shows me that it might be tricky for beginners and would just introduce more annoyance for some people.
> 
> @marko-radosavljevic opinions?

Well I have showed you what happens on my end when I try to do that, if it can be done somewhere in the settings lmk so I can change it than push the changes again

**Comment by marko-radosavljevic** [comment]
@Zabuzard
Yeah, I mostly agree with it, no reason for us to fear compatibility issues.

@IslamSakrak 
The encoding issue you were having is related to our build script. Gradle uses system file encoding when no encoding is specified. This can lead to inconsistent builds, like you noticed. It's a bug we have to fix. ^^

I can write a quick PR to fix this. You can then cherry-pick my commit, and test if it works. Alternatively, I can do this git manipulation for you.

**Comment by IslamSakrak** [comment]
> I can write a quick PR to fix this. You can then cherry-pick my commit, and test if it works. Alternatively, I can do this git manipulation for you.

Sure il wait than something else to work on until i can pull the changes related to the configs so that it gets resolved

**Comment by marko-radosavljevic** [comment]
It's already there, #295. Tell me if you need any help, and tell me if it actually fixes your problem, I haven't tested it. ^^

**Comment by Zabuzard** [comment]
Decided to go with the bullet point inside the source file, hence blocked by #295 .

**Review by Zabuzard — APPROVED** [review]

**Comment by Zabuzard** [comment]
> It's already there, #295. Tell me if you need any help, and tell me if it actually fixes your problem, I haven't tested it. ^^

OP said in Discord it works like a charm 👍

**Comment by Zabuzard** [comment]
Unblocked, thing got merged. Please rebase and test whether it still works as expected. @IslamSakrak

**Comment by IslamSakrak** [comment]
Rebased, and everything still works

**Comment by Zabuzard** [comment]
Perfect. Get yourself another review and it can be merged 👍 (Alternatively, Ill merge it after 7 days of inactivity.)

**Comment by Zabuzard** [comment]
Could you share an up-to-date example screenshot of how it looks? And add it to the PR description please, thanks.

**Comment by IslamSakrak** [comment]
> Could you share an up-to-date example screenshot of how it looks? And add it to the PR description please, thanks.

image

**Comment by Zabuzard** [comment]
Last commit was 7 days ago and it has one approval. Merging.

**Pull request merged**
